refactor(options): reduce `findoption()` usage

famiu opened the pull request:
Follow-up to #26429.

Problem: Many places in the code use `findoption()` to access an option using its name, even if the option index is available. This is very slow because it requires looping through the options array over and over.

Solution: Use option index instead of name wherever possible. Also introduce an `OptIndex` enum which contains the index for every option as enum constants, this eliminates the need to pass static option names as strings.

famiu committed:
refactor(options): reduce `findoption()` usage

Problem: Many places in the code use `findoption()` to access an option using its name, even if the option index is available. This is very slow because it requires looping through the options array over and over.

Solution: Use option index instead of name wherever possible. Also introduce an `OptIndex` enum which contains the index for every option as enum constants, this eliminates the need to pass static option names as strings.

famiu committed:
refactor(options): convert `opt_idx` variables to `OptIndex`

famiu commented on src/nvim/generators/gen_options.lua:
```diff
@@ -229,4 +236,15 @@ w('')
+w('typedef enum {')
+w('  kOptionInvalid = -1,')
+for i, o in ipairs(options.options) do
+  w(('  kOption%s = %u,'):format(title_case(o.full_name), i - 1))
```
Should also probably replace `OptionIndex` with `OptIndex`

zeertzjq commented:
You misspelled the function name.

zeertzjq commented on src/nvim/option.c:
```diff
@@ -3482,7 +3466,7 @@ OptVal get_option_value(int opt_idx, int scope)
-vimoption_T *get_option(int opt_idx)
+vimoption_T *get_option(OptIndex opt_idx)
 {
   return &options[opt_idx];
```
It seems that an assertion is also needed here.

zeertzjq commented on src/nvim/option.c:
```diff
@@ -1700,33 +1689,33 @@ static void didset_options2(void)
 void check_options(void)
 {
-  for (int opt_idx = 0; options[opt_idx].fullname != NULL; opt_idx++) {
+  for (OptIndex opt_idx = 0; options[opt_idx].fullname != NULL; opt_idx++) {
```
I think this change should be reverted, because `OptIndex` doesn't include the index of the final `NULL` option.

famiu replied:
@zeertzjq Instead of reverting it, I instead added a `kOptIndexCount` and removed the `NULL` option. That way we can use `kOptIndexCount` to iterate through the `options[]` array instead.

zeertzjq commented on src/nvim/option.c:
```diff
@@ -4851,19 +4846,19 @@ void didset_window_options(win_T *wp, bool valid_cursor)
   }
   did_init_buf_opt_idx = true;
-  for (int i = 0; options[i].fullname != NULL; i++) {
+  for (OptIndex i = 0; options[i].fullname != NULL; i++) {
```
ditto

zeertzjq commented on src/nvim/option.c:
```diff
@@ -3054,10 +3038,10 @@ int findoption_len(const char *const arg, const size_t len)
   }
 
-  int opt_idx;
+  OptIndex opt_idx;
```
ditto

zeertzjq commented on src/nvim/option.c:
```diff
@@ -288,7 +288,7 @@ static void set_init_default_cdpath(void)
 static void set_init_expand_env(void)
 {
-  for (int opt_idx = 0; options[opt_idx].fullname; opt_idx++) {
+  for (OptIndex opt_idx = 0; options[opt_idx].fullname; opt_idx++) {
```
ditto

zeertzjq commented on src/nvim/option.c:
```diff
@@ -4005,7 +4000,7 @@ static int optval_default(vimoption_T *p, const void *varp)
 void ui_refresh_options(void)
 {
-  for (int opt_idx = 0; options[opt_idx].fullname; opt_idx++) {
+  for (OptIndex opt_idx = 0; options[opt_idx].fullname; opt_idx++) {
```
ditto

zeertzjq commented on src/nvim/option.c:
```diff
@@ -6158,7 +6132,7 @@ dict_T *get_winbuf_options(const int bufopt)
   dict_T *const d = tv_dict_alloc();
 
-  for (int opt_idx = 0; options[opt_idx].fullname; opt_idx++) {
+  for (OptIndex opt_idx = 0; options[opt_idx].fullname; opt_idx++) {
```
ditto

famiu committed:
refactor(options): define `kOptIndexCount`

Add a macro to indicate the option count so that we can iterate through the options[] table more clearly. This also removes the need for having an option with NULL fullname at the end of `options[]`.

bfredl commented on src/nvim/generators/gen_options.lua:
```diff
@@ -41,6 +42,12 @@ local list_flags = {
+--- @param s string
+--- @return string
+local title_case = function(s)
+  return s:sub(1, 1):upper() .. s:sub(2):lower()
```
Isn't all options lowercase anyway so only the `:upper()` part is needed here? Also, if we ever were to add cased option names (which we will not, but if we would), it would be fine to hypothetically preserve that.

bfredl reviewed (approved):
Nice refactor! Needing to look up things stringly internally is always weird.

justinmk commented:
Are these new coverity issues related to this PR (or the other options refactors)? Even if they are false positives, there's an increasing number of them so hopefully we can satisfy coverity via a small code change.

```
8 new defect(s) introduced to neovim/neovim found with Coverity Scan.

*** CID 470408:  Memory - illegal accesses  (STRING_NULL)
/src/nvim/ex_cmds.c: 4668 in show_sub()
4662         }
4663         bufhl_add_hl_pos_offset(orig_buf, cmdpreview_ns, hl_id, match.start, match.end, 0);
4664       }
4665     
4666       xfree(str);
4667     
>>>     CID 470408:  Memory - illegal accesses  (STRING_NULL)
>>>     Passing unterminated string "save_shm_p" to "set_string_option_direct", which expects a null-terminated string.
4668       set_string_option_direct(kOptShortmess, save_shm_p, OPT_FREE, SID_NONE);
4669       xfree(save_shm_p);
4670     
4671       return preview ? 2 : 1;
4672     }
4673     

** CID 470407:  Memory - illegal accesses  (STRING_NULL)


________________________________________________________________________________________________________
*** CID 470407:  Memory - illegal accesses  (STRING_NULL)
/src/nvim/ex_getln.c: 908 in command_line_enter()
902     
903       // When the command line was typed, no need for a wait-return prompt.
904       if (s->some_key_typed && tl_ret) {
905         need_wait_return = false;
906       }
907     
>>>     CID 470407:  Memory - illegal accesses  (STRING_NULL)
>>>     Passing unterminated string "s->save_p_icm" to "set_string_option_direct", which expects a null-terminated string.
908       set_string_option_direct(kOptInccommand, s->save_p_icm, OPT_FREE, SID_NONE);
909       State = s->save_State;
910       if (cmdpreview != save_cmdpreview) {
911         cmdpreview = save_cmdpreview;  // restore preview state
912         redraw_all_later(UPD_SOME_VALID);
913       }

** CID 470406:  Control flow issues  (DEADCODE)
/src/nvim/option.c: 277 in set_init_default_cdpath()


________________________________________________________________________________________________________
*** CID 470406:  Control flow issues  (DEADCODE)
/src/nvim/option.c: 277 in set_init_default_cdpath()
271       buf[j] = NUL;
272       OptIndex opt_idx = kOptCdpath;
273       if (opt_idx != kOptInvalid) {
274         options[opt_idx].def_val = buf;
275         options[opt_idx].flags |= P_DEF_ALLOCED;
276       } else {
>>>     CID 470406:  Control flow issues  (DEADCODE)
>>>     Execution cannot reach this statement: "xfree(buf);".
277         xfree(buf);           // cannot happen
278       }
279       xfree(cdpath);
280     }
281     
282     /// Expand environment variables and things like "~" for the defaults.

** CID 470405:  Memory - illegal accesses  (STRING_NULL)


________________________________________________________________________________________________________
*** CID 470405:  Memory - illegal accesses  (STRING_NULL)
/src/nvim/window.c: 544 in do_window()
538     
539           // If 'switchbuf' is set to 'useopen' or 'usetab' and the
540           // file is already opened in a window, then jump to it.
541           win_T *wp = NULL;
542           if ((swb_flags & (SWB_USEOPEN | SWB_USETAB))
543               && cmdmod.cmod_tab == 0) {
>>>     CID 470405:  Memory - illegal accesses  (STRING_NULL)
>>>     Passing unterminated string "ptr" to "buflist_findname_exp", which expects a null-terminated string.
544             wp = swbuf_goto_win_with_buf(buflist_findname_exp(ptr));
545           }
546     
547           if (wp == NULL && win_split(0, 0) == OK) {
548             RESET_BINDING(curwin);
549             if (do_ecmd(0, ptr, NULL, NULL, ECMD_LASTL, ECMD_HIDE, NULL) == FAIL) {

** CID 470404:  Memory - illegal accesses  (STRING_NULL)


________________________________________________________________________________________________________
*** CID 470404:  Memory - illegal accesses  (STRING_NULL)
/src/nvim/quickfix.c: 5219 in vgr_load_dummy_buf()
5213     
5214       // Load file into a buffer, so that 'fileencoding' is detected,
5215       // autocommands applied, etc.
5216       buf_T *buf = load_dummy_buffer(fname, dirname_start, dirname_now);
5217     
5218       p_mls = save_mls;
>>>     CID 470404:  Memory - illegal accesses  (STRING_NULL)
>>>     Passing unterminated string "save_ei" to "au_event_restore", which expects a null-terminated string.
5219       au_event_restore(save_ei);
5220     
5221       return buf;
5222     }
5223     
5224     /// Check whether a quickfix/location list is valid. Autocmds may remove or

** CID 470403:  Memory - illegal accesses  (STRING_NULL)


________________________________________________________________________________________________________
*** CID 470403:  Memory - illegal accesses  (STRING_NULL)
/src/nvim/ex_docmd.c: 2341 in do_one_cmd()
2335         emsg(errormsg);
2336       }
2337       do_errthrow(cstack,
2338                   (ea.cmdidx != CMD_SIZE
2339                    && !IS_USER_CMDIDX(ea.cmdidx)) ? cmdnames[(int)ea.cmdidx].cmd_name : NULL);
2340     
>>>     CID 470403:  Memory - illegal accesses  (STRING_NULL)
>>>     Passing unterminated string "cmdmod.cmod_save_ei" to "undo_cmdmod", which expects a null-terminated string.
2341       undo_cmdmod(&cmdmod);
2342       cmdmod = save_cmdmod;
2343       reg_executing = save_reg_executing;
2344       pending_end_reg_executing = save_pending_end_reg_executing;
2345     
2346       if (ea.nextcmd && *ea.nextcmd == NUL) {       // not really a next command

** CID 470402:  Memory - illegal accesses  (STRING_NULL)


________________________________________________________________________________________________________
*** CID 470402:  Memory - illegal accesses  (STRING_NULL)
/src/nvim/ex_docmd.c: 1725 in execute_cmd()
1719     
1720     end:
1721       if (errormsg != NULL && *errormsg != NUL) {
1722         emsg(errormsg);
1723       }
1724       // Undo command modifiers
>>>     CID 470402:  Memory - illegal accesses  (STRING_NULL)
>>>     Passing unterminated string "cmdmod.cmod_save_ei" to "undo_cmdmod", which expects a null-terminated string.
1725       undo_cmdmod(&cmdmod);
1726       cmdmod = save_cmdmod;
1727       return retv;
1728     #undef ERROR
1729     }
1730     

** CID 470401:  Memory - illegal accesses  (STRING_NULL)


________________________________________________________________________________________________________
*** CID 470401:  Memory - illegal accesses  (STRING_NULL)
/src/nvim/fileio.c: 1968 in set_forced_fenc()
1962     {
1963       if (eap->force_enc == 0) {
1964         return;
1965       }
1966     
1967       char *fenc = enc_canonize(eap->cmd + eap->force_enc);
>>>     CID 470401:  Memory - illegal accesses  (STRING_NULL)
>>>     Passing unterminated string "fenc" to "set_string_option_direct", which expects a null-terminated string.
1968       set_string_option_direct(kOptFileencoding, fenc, OPT_FREE|OPT_LOCAL, 0);
1969       xfree(fenc);
1970     }
1971     
1972     /// Find next fileencoding to use from 'fileencodings'.
1973     /// "pp" points to fenc_next.  It's advanced to the next item.
```

justinmk commented:
Are these new coverity issues related to this PR (or the other options refactors)? Even if they are false positives, there's an increasing number of them so hopefully we can satisfy coverity via a small code change.

```
8 new defect(s) introduced to neovim/neovim found with Coverity Scan.

*** CID 470408:  Memory - illegal accesses  (STRING_NULL)
/src/nvim/ex_cmds.c: 4668 in show_sub()
4662         }
4663         bufhl_add_hl_pos_offset(orig_buf, cmdpreview_ns, hl_id, match.start, match.end, 0);
4664       }
4665     
4666       xfree(str);
4667     
>>>     CID 470408:  Memory - illegal accesses  (STRING_NULL)
>>>     Passing unterminated string "save_shm_p" to "set_string_option_direct", which expects a null-terminated string.
4668       set_string_option_direct(kOptShortmess, save_shm_p, OPT_FREE, SID_NONE);
4669       xfree(save_shm_p);
4670     
4671       return preview ? 2 : 1;
4672     }
4673     

** CID 470407:  Memory - illegal accesses  (STRING_NULL)


________________________________________________________________________________________________________
*** CID 470407:  Memory - illegal accesses  (STRING_NULL)
/src/nvim/ex_getln.c: 908 in command_line_enter()
902     
903       // When the command line was typed, no need for a wait-return prompt.
904       if (s->some_key_typed && tl_ret) {
905         need_wait_return = false;
906       }
907     
>>>     CID 470407:  Memory - illegal accesses  (STRING_NULL)
>>>     Passing unterminated string "s->save_p_icm" to "set_string_option_direct", which expects a null-terminated string.
908       set_string_option_direct(kOptInccommand, s->save_p_icm, OPT_FREE, SID_NONE);
909       State = s->save_State;
910       if (cmdpreview != save_cmdpreview) {
911         cmdpreview = save_cmdpreview;  // restore preview state
912         redraw_all_later(UPD_SOME_VALID);
913       }

** CID 470406:  Control flow issues  (DEADCODE)
/src/nvim/option.c: 277 in set_init_default_cdpath()


________________________________________________________________________________________________________
*** CID 470406:  Control flow issues  (DEADCODE)
/src/nvim/option.c: 277 in set_init_default_cdpath()
271       buf[j] = NUL;
272       OptIndex opt_idx = kOptCdpath;
273       if (opt_idx != kOptInvalid) {
274         options[opt_idx].def_val = buf;
275         options[opt_idx].flags |= P_DEF_ALLOCED;
276       } else {
>>>     CID 470406:  Control flow issues  (DEADCODE)
>>>     Execution cannot reach this statement: "xfree(buf);".
277         xfree(buf);           // cannot happen
278       }
279       xfree(cdpath);
280     }
281     
282     /// Expand environment variables and things like "~" for the defaults.

** CID 470405:  Memory - illegal accesses  (STRING_NULL)


________________________________________________________________________________________________________
*** CID 470405:  Memory - illegal accesses  (STRING_NULL)
/src/nvim/window.c: 544 in do_window()
538     
539           // If 'switchbuf' is set to 'useopen' or 'usetab' and the
540           // file is already opened in a window, then jump to it.
541           win_T *wp = NULL;
542           if ((swb_flags & (SWB_USEOPEN | SWB_USETAB))
543               && cmdmod.cmod_tab == 0) {
>>>     CID 470405:  Memory - illegal accesses  (STRING_NULL)
>>>     Passing unterminated string "ptr" to "buflist_findname_exp", which expects a null-terminated string.
544             wp = swbuf_goto_win_with_buf(buflist_findname_exp(ptr));
545           }
546     
547           if (wp == NULL && win_split(0, 0) == OK) {
548             RESET_BINDING(curwin);
549             if (do_ecmd(0, ptr, NULL, NULL, ECMD_LASTL, ECMD_HIDE, NULL) == FAIL) {

** CID 470404:  Memory - illegal accesses  (STRING_NULL)


________________________________________________________________________________________________________
*** CID 470404:  Memory - illegal accesses  (STRING_NULL)
/src/nvim/quickfix.c: 5219 in vgr_load_dummy_buf()
5213     
5214       // Load file into a buffer, so that 'fileencoding' is detected,
5215       // autocommands applied, etc.
5216       buf_T *buf = load_dummy_buffer(fname, dirname_start, dirname_now);
5217     
5218       p_mls = save_mls;
>>>     CID 470404:  Memory - illegal accesses  (STRING_NULL)
>>>     Passing unterminated string "save_ei" to "au_event_restore", which expects a null-terminated string.
5219       au_event_restore(save_ei);
5220     
5221       return buf;
5222     }
5223     
5224     /// Check whether a quickfix/location list is valid. Autocmds may remove or

** CID 470403:  Memory - illegal accesses  (STRING_NULL)


________________________________________________________________________________________________________
*** CID 470403:  Memory - illegal accesses  (STRING_NULL)
/src/nvim/ex_docmd.c: 2341 in do_one_cmd()
2335         emsg(errormsg);
2336       }
2337       do_errthrow(cstack,
2338                   (ea.cmdidx != CMD_SIZE
2339                    && !IS_USER_CMDIDX(ea.cmdidx)) ? cmdnames[(int)ea.cmdidx].cmd_name : NULL);
2340     
>>>     CID 470403:  Memory - illegal accesses  (STRING_NULL)
>>>     Passing unterminated string "cmdmod.cmod_save_ei" to "undo_cmdmod", which expects a null-terminated string.
2341       undo_cmdmod(&cmdmod);
2342       cmdmod = save_cmdmod;
2343       reg_executing = save_reg_executing;
2344       pending_end_reg_executing = save_pending_end_reg_executing;
2345     
2346       if (ea.nextcmd && *ea.nextcmd == NUL) {       // not really a next command

** CID 470402:  Memory - illegal accesses  (STRING_NULL)


________________________________________________________________________________________________________
*** CID 470402:  Memory - illegal accesses  (STRING_NULL)
/src/nvim/ex_docmd.c: 1725 in execute_cmd()
1719     
1720     end:
1721       if (errormsg != NULL && *errormsg != NUL) {
1722         emsg(errormsg);
1723       }
1724       // Undo command modifiers
>>>     CID 470402:  Memory - illegal accesses  (STRING_NULL)
>>>     Passing unterminated string "cmdmod.cmod_save_ei" to "undo_cmdmod", which expects a null-terminated string.
1725       undo_cmdmod(&cmdmod);
1726       cmdmod = save_cmdmod;
1727       return retv;
1728     #undef ERROR
1729     }
1730     

** CID 470401:  Memory - illegal accesses  (STRING_NULL)


________________________________________________________________________________________________________
*** CID 470401:  Memory - illegal accesses  (STRING_NULL)
/src/nvim/fileio.c: 1968 in set_forced_fenc()
1962     {
1963       if (eap->force_enc == 0) {
1964         return;
1965       }
1966     
1967       char *fenc = enc_canonize(eap->cmd + eap->force_enc);
>>>     CID 470401:  Memory - illegal accesses  (STRING_NULL)
>>>     Passing unterminated string "fenc" to "set_string_option_direct", which expects a null-terminated string.
1968       set_string_option_direct(kOptFileencoding, fenc, OPT_FREE|OPT_LOCAL, 0);
1969       xfree(fenc);
1970     }
1971     
1972     /// Find next fileencoding to use from 'fileencodings'.
1973     /// "pp" points to fenc_next.  It's advanced to the next item.
```

famiu commented:
> Are these new coverity issues related to this PR (or the other options refactors)? Even if they are false positives, there's an increasing number of them so hopefully we can satisfy coverity via a small code change.
> 
> ```
> 8 new defect(s) introduced to neovim/neovim found with Coverity Scan.
> 
> *** CID 470408:  Memory - illegal accesses  (STRING_NULL)
> /src/nvim/ex_cmds.c: 4668 in show_sub()
> 4662         }
> 4663         bufhl_add_hl_pos_offset(orig_buf, cmdpreview_ns, hl_id, match.start, match.end, 0);
> 4664       }
> 4665     
> 4666       xfree(str);
> 4667     
> >>>     CID 470408:  Memory - illegal accesses  (STRING_NULL)
> >>>     Passing unterminated string "save_shm_p" to "set_string_option_direct", which expects a null-terminated string.
> 4668       set_string_option_direct(kOptShortmess, save_shm_p, OPT_FREE, SID_NONE);
> 4669       xfree(save_shm_p);
> 4670     
> 4671       return preview ? 2 : 1;
> 4672     }
> 4673     
> 
> ** CID 470407:  Memory - illegal accesses  (STRING_NULL)
> 
> 
> ________________________________________________________________________________________________________
> *** CID 470407:  Memory - illegal accesses  (STRING_NULL)
> /src/nvim/ex_getln.c: 908 in command_line_enter()
> 902     
> 903       // When the command line was typed, no need for a wait-return prompt.
> 904       if (s->some_key_typed && tl_ret) {
> 905         need_wait_return = false;
> 906       }
> 907     
> >>>     CID 470407:  Memory - illegal accesses  (STRING_NULL)
> >>>     Passing unterminated string "s->save_p_icm" to "set_string_option_direct", which expects a null-terminated string.
> 908       set_string_option_direct(kOptInccommand, s->save_p_icm, OPT_FREE, SID_NONE);
> 909       State = s->save_State;
> 910       if (cmdpreview != save_cmdpreview) {
> 911         cmdpreview = save_cmdpreview;  // restore preview state
> 912         redraw_all_later(UPD_SOME_VALID);
> 913       }
> 
> ** CID 470406:  Control flow issues  (DEADCODE)
> /src/nvim/option.c: 277 in set_init_default_cdpath()
> 
> 
> ________________________________________________________________________________________________________
> *** CID 470406:  Control flow issues  (DEADCODE)
> /src/nvim/option.c: 277 in set_init_default_cdpath()
> 271       buf[j] = NUL;
> 272       OptIndex opt_idx = kOptCdpath;
> 273       if (opt_idx != kOptInvalid) {
> 274         options[opt_idx].def_val = buf;
> 275         options[opt_idx].flags |= P_DEF_ALLOCED;
> 276       } else {
> >>>     CID 470406:  Control flow issues  (DEADCODE)
> >>>     Execution cannot reach this statement: "xfree(buf);".
> 277         xfree(buf);           // cannot happen
> 278       }
> 279       xfree(cdpath);
> 280     }
> 281     
> 282     /// Expand environment variables and things like "~" for the defaults.
> 
> ** CID 470405:  Memory - illegal accesses  (STRING_NULL)
> 
> 
> ________________________________________________________________________________________________________
> *** CID 470405:  Memory - illegal accesses  (STRING_NULL)
> /src/nvim/window.c: 544 in do_window()
> 538     
> 539           // If 'switchbuf' is set to 'useopen' or 'usetab' and the
> 540           // file is already opened in a window, then jump to it.
> 541           win_T *wp = NULL;
> 542           if ((swb_flags & (SWB_USEOPEN | SWB_USETAB))
> 543               && cmdmod.cmod_tab == 0) {
> >>>     CID 470405:  Memory - illegal accesses  (STRING_NULL)
> >>>     Passing unterminated string "ptr" to "buflist_findname_exp", which expects a null-terminated string.
> 544             wp = swbuf_goto_win_with_buf(buflist_findname_exp(ptr));
> 545           }
> 546     
> 547           if (wp == NULL && win_split(0, 0) == OK) {
> 548             RESET_BINDING(curwin);
> 549             if (do_ecmd(0, ptr, NULL, NULL, ECMD_LASTL, ECMD_HIDE, NULL) == FAIL) {
> 
> ** CID 470404:  Memory - illegal accesses  (STRING_NULL)
> 
> 
> ________________________________________________________________________________________________________
> *** CID 470404:  Memory - illegal accesses  (STRING_NULL)
> /src/nvim/quickfix.c: 5219 in vgr_load_dummy_buf()
> 5213     
> 5214       // Load file into a buffer, so that 'fileencoding' is detected,
> 5215       // autocommands applied, etc.
> 5216       buf_T *buf = load_dummy_buffer(fname, dirname_start, dirname_now);
> 5217     
> 5218       p_mls = save_mls;
> >>>     CID 470404:  Memory - illegal accesses  (STRING_NULL)
> >>>     Passing unterminated string "save_ei" to "au_event_restore", which expects a null-terminated string.
> 5219       au_event_restore(save_ei);
> 5220     
> 5221       return buf;
> 5222     }
> 5223     
> 5224     /// Check whether a quickfix/location list is valid. Autocmds may remove or
> 
> ** CID 470403:  Memory - illegal accesses  (STRING_NULL)
> 
> 
> ________________________________________________________________________________________________________
> *** CID 470403:  Memory - illegal accesses  (STRING_NULL)
> /src/nvim/ex_docmd.c: 2341 in do_one_cmd()
> 2335         emsg(errormsg);
> 2336       }
> 2337       do_errthrow(cstack,
> 2338                   (ea.cmdidx != CMD_SIZE
> 2339                    && !IS_USER_CMDIDX(ea.cmdidx)) ? cmdnames[(int)ea.cmdidx].cmd_name : NULL);
> 2340     
> >>>     CID 470403:  Memory - illegal accesses  (STRING_NULL)
> >>>     Passing unterminated string "cmdmod.cmod_save_ei" to "undo_cmdmod", which expects a null-terminated string.
> 2341       undo_cmdmod(&cmdmod);
> 2342       cmdmod = save_cmdmod;
> 2343       reg_executing = save_reg_executing;
> 2344       pending_end_reg_executing = save_pending_end_reg_executing;
> 2345     
> 2346       if (ea.nextcmd && *ea.nextcmd == NUL) {       // not really a next command
> 
> ** CID 470402:  Memory - illegal accesses  (STRING_NULL)
> 
> 
> ________________________________________________________________________________________________________
> *** CID 470402:  Memory - illegal accesses  (STRING_NULL)
> /src/nvim/ex_docmd.c: 1725 in execute_cmd()
> 1719     
> 1720     end:
> 1721       if (errormsg != NULL && *errormsg != NUL) {
> 1722         emsg(errormsg);
> 1723       }
> 1724       // Undo command modifiers
> >>>     CID 470402:  Memory - illegal accesses  (STRING_NULL)
> >>>     Passing unterminated string "cmdmod.cmod_save_ei" to "undo_cmdmod", which expects a null-terminated string.
> 1725       undo_cmdmod(&cmdmod);
> 1726       cmdmod = save_cmdmod;
> 1727       return retv;
> 1728     #undef ERROR
> 1729     }
> 1730     
> 
> ** CID 470401:  Memory - illegal accesses  (STRING_NULL)
> 
> 
> ________________________________________________________________________________________________________
> *** CID 470401:  Memory - illegal accesses  (STRING_NULL)
> /src/nvim/fileio.c: 1968 in set_forced_fenc()
> 1962     {
> 1963       if (eap->force_enc == 0) {
> 1964         return;
> 1965       }
> 1966     
> 1967       char *fenc = enc_canonize(eap->cmd + eap->force_enc);
> >>>     CID 470401:  Memory - illegal accesses  (STRING_NULL)
> >>>     Passing unterminated string "fenc" to "set_string_option_direct", which expects a null-terminated string.
> 1968       set_string_option_direct(kOptFileencoding, fenc, OPT_FREE|OPT_LOCAL, 0);
> 1969       xfree(fenc);
> 1970     }
> 1971     
> 1972     /// Find next fileencoding to use from 'fileencodings'.
> 1973     /// "pp" points to fenc_next.  It's advanced to the next item.
> ```

Doesn't seem related to the PR
